pkg/unpack/tag.go

```
@@ -0,0 +1,100 @@
+	s, ok, _ := parseTag(tagJSON, f)
+	return s, ok
+}
+
+// structToRule looks through the tags in a struct to (1) make sure there
+// are no duplicate tags since package json doesn't do this and it results
+// in hard-to-debug behaviors, and (2) it finds an "unpack" tag which indicates
+// which field of the struct to use as the key and what value to match for
+// unmarshling into this concrete type.  If there is not exactly one unpack
+// tag, then an error is returned.
+// unpack="key,skip" unpack="key", unpack="", or unpack="skip"
+func structToUnpackRule(typ reflect.Type) (string, string, bool, error) {
+	if typ.Kind() != reflect.Struct {
+		return "", "", false, errors.New("cannot unpack into non-struct")
+	}
+	names := make(map[string]struct{})
+	var unpackKey string
+	var unpackVal string
+	var unpackSkip bool
+	for k := 0; k < typ.NumField(); k++ {
+		field := typ.Field(k)
+		jsonField, jsonOk, _ := parseTag(tagJSON, field)
+		if jsonOk {
+			if _, ok := names[jsonField]; ok {
+				return "", "", false, fmt.Errorf("json field tag '%s' in struct type '%s' not unique", jsonField, typ.Name())
+			}
+			names[jsonField] = struct{}{}
+		}
+		unpackOpt, unpackOk, opts := parseTag(tagUnpack, typ.Field(k))
+		if !unpackOk {
+			continue
+		}
+		if len(opts) > 1 {
+			return "", "", false, fmt.Errorf("unpack: too many tag options in field '%s' of struct type '%s'", field.Name, typ.Name())
+		}
+		var skip bool
+		if len(opts) == 1 {
+			if opts[0] != "skip" {
+				return "", "", false, fmt.Errorf("unpack: second tag option in field '%s' of struct type '%s' may only be 'skip'", field.Name, typ.Name())
+			}
+			skip = true
+		}
+		if skip {
+			if unpackSkip {
+				return "", "", false, ErrSkip
+			}
+			unpackSkip = true
+		}
+		if unpackKey != "" {
+			return "", "", false, fmt.Errorf("unpack key appears twice in field '%s' of struct type '%s' may only be 'skip'", field.Name, typ.Name())
+		}
+		if jsonField == "" {
```

Might want to return an error if the JSON field name is `-`.